src/ponyfills/objects.js

```diff
@@ -0,0 +1,176 @@
+ *
+ * @param {Object} map A map with hierarchical keys
+ * @returns {Object} A map with hierarchical keys flattened
+ */
+export function flatten(map, filter) {
+  function recurse(current, parent_key) {
+    let items = {};
+    Object.keys(current).filter(filter || function() { return true; }).forEach(function(k) {
+      let v = current[k];
+      let new_key = parent_key ? (parent_key + '.' + k) : k;
+      if (isObject(v)) {
+        items = assign(items, recurse(current[k], new_key));
+      } else {
+        items[new_key] = v;
+      }
+    });
+
+    return items;
+  }
+
+  return recurse(map, '');
+}
+
+export function flattenKeys(map, filter) {
+  function recurse(current, parent_key) {
+    let items = [];
+    Object.keys(current).filter(filter || function() { return true; }).forEach(function(k) {
+      let v = current[k];
+      let new_key = parent_key ? (parent_key + '.' + k) : k;
+      items.push(new_key);
+      if (isObject(v)) {
+        items = items.concat(recurse(current[k], new_key));
+      }
+    });
+
+    return items;
+  }
+
+  return recurse(map, '');
+}
+
+export function removeValueForKey(key, map) {
+  function recurse(keys, index, map) {
+    let key = keys[index];
+    if (index === (keys.length - 1)) {
+      delete map[key];
+      return true;
+    }
+
+    if (!(key in map)) {
+      return false;
+    }
+
+    if (recurse(keys, index + 1, map[key]) && Object.keys(map[key]).length === 0) {
+      delete map[key];
+      return true;
+    }
+  }
+
+  return recurse(key.split('\.'), 0, map);
+}
+
+export function getValueForKey(key, map) {
+  let value;
+  let current = map;
+  let keys = key.split('\.');
+  for (let i = 0; i < keys.length; i++) {
+    let k = keys[i];
+    if (i === (keys.length - 1)) {
+      value = current[k];
+      break;
+    }
+
+    if (!(k in current)) {
+      break;
+    }
+
+    current = current[k];
+  }
+
+  return value;
+}
+
+export function setKeyToValue(key, value, map) {
+  let current = map;
+  let keys = key.split('\.');
+  for (let i = 0; i < keys.length; i++) {
+    let k = keys[i];
+    if (i === (keys.length - 1)) {
+      current[k] = value;
+      break;
+    }
+
+    if (!(k in current)) {
+      current[k] = {};
+    }
+
+    current = current[k];
+  }
+
+  return value;
+}
+
+/**
+ * Convert a flattened map into a hierarchical map
+ *
+ * @param {Object} map A map with hierarchical keys flattened
+ * @returns {Object} A map with hierarchical keys
+ */
+export function expand(map) {
+  let expanded = {};
+  Object.keys(map).forEach(function(key) {
+    let v = map[key];
+    setKeyToValue(key, v, expanded);
+  });
+
+  return expanded;
+}
+
+export function filter(map, active) {
+  const flattened = flatten(map);
+  const filtered = {};
+  active.forEach(function(key) {
+    if (key in flattened) {
+      filtered[key] = flattened[key];
+    }
+  });
+
+  return expand(filtered);
+}
+
+export function assign(target, sources) {
```

Comment:
Following my earlier recommendation for transpiling, a transpiler would ponyfill this method automatically, relieving contributors from needing to know to use this special method.

Reply:
This library is implemented to ensure awareness of the impacts of code changes. I'm specifically trying to ensure that a contributor doesn't make a minor change that doubles the size of the library without them knowing. This will be a downstream dependency of most of what we build in the JS stack, so we need to be really thoughtful on introducing size and complexity only when it adds significant value.

Reply:
I'm sympathetic to that too (that is, of adding bulk). A transpiler isn't going to any more bulk than this function is going to. In fact, it would remove weight in the instances where the SDK is used as a Node library since you're already insisting that the engine be Node 13. There are also weight-shedding optimizations that can be made for evergreen browsers too with differential serving.

I've got to be honest: I think avoiding a transpiler adds more cognitive complexity than including one. It exchanges complexity in build steps for extra complexity in writing code, reading code, and ensuring that the JS you're using is ES5 compatible.